### PR TITLE
Bug 1964540: Extend trunk configuration to port level in machineset

### PR DESCRIPTION
**Pull request by adduarte**
Extend trunk configuration to port level in machineset
With the introduction of new technologies, like SR-IOV,
it is required that the user have the ability to
configured trunked and not-trunked ports for the same machines
using spec.ports.trunk.

- Extended trunk support needed check to cover spec.port and spec.network
- Makes ports inherit trunk setting from spec.Trunk if not specified.
- Refactored trunking to its own function. Trunk name matches port name
- Added trunking to spec.ports

[OSASINFRA-2504](https://issues.redhat.com/browse/OSASINFRA-2504)

### CODE REVIEW

**Comment by adduarte** [comment]
As it currently stands, this patch brings us up to par with kubernetes-sigs/cluster-api-provider-openstack functionality: 
you can define interfaces under "network" and "port"  and you can trunk All of them or None of them. 

This make a deployment using kuryr  but with SR-IOV ports not trunked impossible.

**Comment by adduarte** [comment]
This functionality is being addressed upstream in pr https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/934. 
Is best if we rework this pr to mirror upstream code and logic.

**Comment by dulek** [comment]
Looks good to me, but I'll leave the `lgtm` label for folks that are more familiar with this code.

File: pkg/cloud/openstack/clients/machineservice.go [hunk 1]
```diff
@@ -458,14 +458,62 @@ func GetSecurityGroups(is *InstanceService, sg_param []openstackconfigv1.Securit
...
+	return false
+}
+
+func getOrCreateTrunk(is *InstanceService, port *ports.Port, machineTags []string) (*trunks.Trunk, error) {
+	allPages, err := trunks.List(is.networkClient, trunks.ListOpts{
+		Name:   port.Name,
+		PortID: port.ID,
+	}).AllPages()
+	if err != nil {
+		return nil, fmt.Errorf("Searching for existing trunk for server err: %v", err)
```
Comment (mandre) [inline pkg/cloud/openstack/clients/machineservice.go:479]: Please update the error message to include the port id, such as:
```suggestion
		return nil, fmt.Errorf("Searching for existing trunk for port %q err: %v", port.ID, err)
```

File: pkg/cloud/openstack/clients/machineservice.go [hunk 2]
```diff
@@ -458,14 +458,62 @@ func GetSecurityGroups(is *InstanceService, sg_param []openstackconfigv1.Securit
...
+	}
+	trunkList, err := trunks.ExtractTrunks(allPages)
+	if err != nil {
+		return nil, fmt.Errorf("Searching for existing trunk for server err: %v", err)
```
Comment (mandre) [inline pkg/cloud/openstack/clients/machineservice.go:483]: ditto

File: pkg/cloud/openstack/clients/machineservice.go [hunk 3]
```diff
@@ -458,14 +458,62 @@ func GetSecurityGroups(is *InstanceService, sg_param []openstackconfigv1.Securit
...
+	var trunk trunks.Trunk
+	if len(trunkList) == 0 {
+		// create trunk with the previous port as parent
+		trunkCreateOpts := trunks.CreateOpts{
+			Name:   port.Name,
+			PortID: port.ID,
+		}
+		newTrunk, err := trunks.Create(is.networkClient, trunkCreateOpts).Extract()
+		if err != nil {
+			return nil, fmt.Errorf("Create trunk for server err: %v", err)
```
Comment (mandre) [inline pkg/cloud/openstack/clients/machineservice.go:494]: ditto

File: pkg/cloud/openstack/clients/machineservice.go [hunk 4]
```diff
@@ -458,14 +458,62 @@ func GetSecurityGroups(is *InstanceService, sg_param []openstackconfigv1.Securit
...
+		}
+		trunk = *newTrunk
+	} else {
+		trunk = trunkList[0]
+	}
+
+	_, err = attributestags.ReplaceAll(is.networkClient, "trunks", trunk.ID, attributestags.ReplaceAllOpts{
+		Tags: machineTags}).Extract()
+	if err != nil {
+		return nil, fmt.Errorf("Tagging trunk for server err: %v", err)
```
Comment (mandre) [inline pkg/cloud/openstack/clients/machineservice.go:504]: ditto

**Review by mandre — COMMENTED** [review]
I _think_ it should work - CI at least shows there is no regression. Were you able to test this patch?

File: pkg/cloud/openstack/clients/machineservice.go (hunk 2 above)
Comment (adduarte) [reply pkg/cloud/openstack/clients/machineservice.go:483]: done

File: pkg/cloud/openstack/clients/machineservice.go (hunk 3 above)
Comment (adduarte) [reply pkg/cloud/openstack/clients/machineservice.go:494]: done

File: pkg/cloud/openstack/clients/machineservice.go (hunk 4 above)
Comment (adduarte) [reply pkg/cloud/openstack/clients/machineservice.go:504]: done

**Comment by adduarte** [comment]
I should have enough testing done locally to verify its functionality soon.

File: pkg/cloud/openstack/clients/machineservice.go [hunk 5]
```diff
@@ -458,14 +458,62 @@ func GetSecurityGroups(is *InstanceService, sg_param []openstackconfigv1.Securit
...
+	var trunk trunks.Trunk
+	if len(trunkList) == 0 {
+		// create trunk with the previous port as parent
+		trunkCreateOpts := trunks.CreateOpts{
+			Name:   port.Name,
+			PortID: port.ID,
+		}
+		newTrunk, err := trunks.Create(is.networkClient, trunkCreateOpts).Extract()
+		if err != nil {
+			return nil, fmt.Errorf("Crete trunk for port %q err: %v", port.ID, err)
```
Comment (mandre) [inline pkg/cloud/openstack/clients/machineservice.go:494]: Nit: Create

Comment (adduarte) [reply pkg/cloud/openstack/clients/machineservice.go]: done

**Comment by adduarte** [comment]
@mandre just spend the day testing and it seems to work as expected. Made some trivial changes to name trunks consistently. 
But I tested 
network.trunk = true & port.trunk=(nil|false,true) and network.trunk=false & port.trunk=(nil,false,true) and all seems to work as expected.

**Comment by adduarte** [comment]
Also even tested with *only*  spec.port, and no spec.network definition. The machine comes up connected to the correct port.

File: pkg/cloud/openstack/clients/machineservice.go [hunk 6]
```diff
@@ -458,14 +457,66 @@ func GetSecurityGroups(is *InstanceService, sg_param []openstackconfigv1.Securit
...
+		}
+	}
+	return false
+}
+
+func getOrCreateTrunk(is *InstanceService, port *ports.Port, machineTags []string) (*trunks.Trunk, error) {
+	trunkName := port.Name + "-trunk"
+	if len(trunkName) > PortNameMaxSize {
+		trunkName = trunkName[len(trunkName)-PortNameMaxSize:]
+	}
```
Comment (mandre) [inline pkg/cloud/openstack/clients/machineservice.go:476]: You're introducing a backward incompatible change. If you want to rename the trunk ports to have the trunk suffix in their names, you should do it in a separate PR, that we first submit upstream.

**Review by mandre — COMMENTED** [review]
LGTM apart from the trunk port renaming. Please revert this change and let's merge the PR.

File: pkg/cloud/openstack/clients/machineservice.go (hunk 6 above)
Comment (adduarte) [reply pkg/cloud/openstack/clients/machineservice.go]: Done. 
I did have to live trunkName := port.Name 
because otherwise we introduce a bug, for example if we have network.trunk and port.trunk...  Then the second port.trunk will  find the trunk for network.trunk. and the port won't get trunked. 
It should not introduce a regression since all ports are created before the trunk and they get a name.

**Comment by adduarte** [comment]
Addressed comment about trunkName. See inline reply

**Comment by adduarte** [comment]
@dulek could you take a ,ook at the code that sets the name of the trunk. 
Perhaps we shoudl not change the name of the trunk at all...

File: pkg/cloud/openstack/clients/machineservice.go (hunk 6 above)
Comment (adduarte) [reply pkg/cloud/openstack/clients/machineservice.go]: actually, changing the name of the trunk will probably have effects down the line. 
@dulek, Does the name of the trun need to match the port, or the hostname?

Comment (adduarte) [reply pkg/cloud/openstack/clients/machineservice.go]: Right now as the patch extends.. only one trunk is created per host, using the name of the host/vm. 

This is in line with 4.8 
For example when someone does: 

networks: 
     - netuuid: primary_network
     - netuuid: someothernetwork
trunk = true


in this case only one trunk is created per host

**Comment by iamemilio** [comment]
we should probably wait for the upstream patch to settle and merge before approving this, it seems like they are still addressing some comments.

**Comment by mandre** [comment]
> we should probably wait for the upstream patch to settle and merge before approving this, it seems like they are still addressing some comments.

Keep in mind we're talking about downstream CAPO and we want to retire this code soon anyway. We don't have to wait for the upstream code to merge (we'll inherit it anyway once we move to MAPO), however we need to make absolutely sure the API is stable and agreed on upstream. This is what we must implement in this patch.

File: pkg/cloud/openstack/clients/machineservice.go (hunk 6 above)
Comment (adduarte) [reply pkg/cloud/openstack/clients/machineservice.go]: As the patch exists right now if we have 
spec.trunk=true
and 
port.trunk=true

two ports are created (same as current upstream). 
If only one trunk needs to be created then user needs to explicitely set all relevant port.trunk=false (or spec.trunk=false)
for example: 
spec.trunk=true
port.trunk= false
and 
spec.trunk=false
port.trunk=true
both result on a single trunk being created. 
spec.trunk = true
and port.trunk=undefined
creates two trunks because port.trunk default is =spec.trunk

**Comment by adduarte** [comment]
the current version of this patch mimics upstream: 
trunks are created with trunkname=port.name,

**Comment by iamemilio** [comment]
> Keep in mind we're talking about downstream CAPO and we want to retire this code soon anyway. We don't have to wait for the upstream code to merge (we'll inherit it anyway once we move to MAPO), however we need to make absolutely sure the API is stable and agreed on upstream. This is what we must implement in this patch.

true, as long as the upstream api is stable, I suppose its ok

File: pkg/cloud/openstack/clients/machineservice.go [hunk 7]
```diff
@@ -330,10 +330,9 @@ func getSubnetsByFilter(is *InstanceService, opts *subnets.ListOpts) ([]subnets.
 
 func getOrCreatePort(is *InstanceService, name string, portOpts openstackconfigv1.PortOpts) (*ports.Port, error) {
 	var portName string
+	portName = name
```
Comment (iamemilio) [inline pkg/cloud/openstack/clients/machineservice.go:333]:
```suggestion
	portName := name
```
nit pick

**Comment by iamemilio** [comment]
little nit pick, but otherwise lgtm
